pointcloud/octree/basic.go

```diff
@@ -0,0 +1,173 @@
+package octree
+
+import (
+	"context"
+
+	"github.com/edaniels/golog"
+	"github.com/golang/geo/r3"
+	"github.com/pkg/errors"
+
+	pc "go.viam.com/rdk/pointcloud"
+)
+
+// basicOctree is a data structure that represents a basic octree structure with information regarding center
+// point, side length and node data. Node data is comprised of a PointAndData (should one exist) from
+//pointcloud, a list of children nodes and the type of node this is. See nodeType constants in octree.go
+type basicOctree struct {
+	ctx    context.Context
+	logger golog.Logger
+	node   basicOctreeNode
+	center r3.Vector
+	side   float64
+	meta   MetaData
+}
+
+type basicOctreeNode struct {
+	nodeType NodeType
+	tree     []*basicOctree
+	point    pc.PointAndData
+}
+
+// New creates a new basic octree with specified center, side and metadata.
+func New(ctx context.Context, center r3.Vector, sideLength float64, logger golog.Logger) (Octree, error) {
+	if sideLength <= 0 {
+		return nil, errors.Errorf("invalid side length (%.2f) for octree", sideLength)
+	}
+
+	octree := &basicOctree{
+		ctx:    ctx,
+		logger: logger,
+		node:   newLeafNodeEmpty(),
+		center: center,
+		side:   sideLength,
+	}
+	octree.meta = octree.NewMetaData()
+	return octree, nil
+}
+
+// Size returns the number of points stored in the octree by traversing it and incrementing based on the number
+// of nodes containing a point.
+func (octree *basicOctree) Size() int {
+	var totalSize int
+
+	switch octree.node.nodeType {
+	case InternalNode:
+		for _, children := range octree.node.tree {
+			totalSize += children.Size()
+		}
+
+	case LeafNodeFilled:
+		totalSize = 1
+
+	case LeafNodeEmpty:
+		totalSize = 0
+	}
+
+	return totalSize
+}
+
+// Set checks if the point to be added is a valid point for the octree to hold based on its center and side length.
+// It then recursively iterates through the tree until it finds the appropriate node to add it too. If the found node
+// contains a point already, it will split the node into octants and add both the old and new points to them respectively.
+func (octree *basicOctree) Set(p r3.Vector, d pc.Data) error {
+	if (pc.PointAndData{P: p, D: d} == pc.PointAndData{}) {
```

Comment:
Will this condition ever happen? Create a test to trigger this condition.

Reply:
It should not happen in nominal operation, however, it is possible for a user to try to set a blank point. If so I believe operation should not error out but it should be logged. I don't have this same check in At as if it is not possible for an empty point to be set, At will always produce the false flag as desired. I will add a test condition for this.

The nil return of this case is tested in the artificial event: split octants of an empty tree. A test for checking that the log was triggered can be added if you think necessary